### ISSUE TITLE
[DISCUSSION] Simplify CocoaLumberjack

### ISSUE THREAD

**Issue by sushichop** [CocoaLumberjack/CocoaLumberjack#954]
### New Issue Checklist

* [x] I have read and understood the [CONTRIBUTING guide](https://github.com/CocoaLumberjack/CocoaLumberjack/blob/master/.github/CONTRIBUTING.md)
* [x] I have read the [Documentation](https://github.com/CocoaLumberjack/CocoaLumberjack#documentation) or [CocoaDocs](cocoadocs.org/docsets/CocoaLumberjack)
* [x] I have searched for a similar issue in the [project](https://github.com/CocoaLumberjack/CocoaLumberjack/issues) and found none

IMHO, the structure of CocoaLumberjack is becoming complicated.
Then, I want to simplify CocoaLumberjack in order to maintain it flexibly and easily for the future.

**2 necessary things to simplify**
1)  I want to change minimum os versions to iOS 8.0, macOS 10.10, tvOS 9.0, and watchOS 2.0. 
These values should be low enough for many developers and, furthermore, these are the same as the minimum versions of most libraries.

2) I want to limit podspecs to only 2 specs. default(Objective-C) subspec and swift subspec.
I think most developers use only these 2 specs. Of course, I'll add the old subspecs(extension and cli subspecs) to the new default and swift subspec.

What do you think about this?
If you agree, I'll be glad to create a PR for it.

---
**Background**

After finishing the above, I want to introduce universal(multi platform) framework instead of each targeted platform framework.
 - CocoaLumberjack- {iOS, macOS, tvOS, watchOS} (4 frameworks) -> CocoaLumberjack (1 framework)
 - CocoaLumberjackSwift- {iOS, macOS, tvOS, watchOS} (4 frameworks) -> CocoaLumberjackSwift (1 framework)

I also think the simplicity as above is necessary as a first step to support SwiftPM in the near future.

Thanks.

**Comment by ffried** [comment]
> @ffried the idea with xcconfig files was to simplify all our projects (since we have a lot in this repo) and to have a central place with these settings. I have used this approach in other projects and it reduces the conflicts, basically the xcodeproj will keep files and targets and the Xcode settings are in the xcconfig files.

Yea, I saw that there are a few projects in the repo. I think we can already improve things a lot by merging these into one project. And I think once we have merged these, there won't be many settings that can't be set on a project level. The universal frameworks I've worked with (e.g. [DDMathParser](https://github.com/davedelong/DDMathParser) or my own [FFFoundation](https://github.com/ffried/FFFoundation)) don't really have conflicts in the project file. It'll basically only change when someone adds a file and two or three times a year for the Xcode / Swift upgrades.
Unless I'm missing something, currently, there are only two sets of settings: One for the frameworks and one for the test apps. Once the frameworks are merged, there's only one framework left, so its settings can be set on the target directly. For the test app it's then probably not much of a difference if they're set on a project level or in an xcconfig file.

> Just a proposal, not an enforcement :)

Don't get me wrong, I'm not voting against xcconfig files here. Maybe I just don't see the advantage in this case (yet). 🙂 But I also don't really care where I change build settings (as long as its clear where I have to change them). xcconfig files or Xcode's build settings editor - either way is fine by me. 🙂

**Comment by ffried** [comment]
(Sorry, wrong button 😬)

**Comment by bpoplauschi** [comment]
@ffried I think it's good to discuss those details :)

We have a lot of Xcode projects (I didn't get to change those in my branch).
```
./Lumberjack.xcodeproj
./Tests/CocoaLumberjack Tests.xcodeproj
./Tests/Pods/Pods.xcodeproj - we can exclude this one
./Demos/Benchmark/Desktop/BenchmarkMac.xcodeproj
./Demos/Benchmark/Mobile/BenchmarkIPhone.xcodeproj
./Demos/SQLiteLogger/SQLiteLogger.xcodeproj
./Demos/FineGrainedLogging/FineGrainedLogging.xcodeproj
./Demos/CustomLogLevels/CustomLogLevels.xcodeproj
./Demos/CustomLogLevels/Pods/Pods.xcodeproj
./Demos/ContextFilter/ContextFilter.xcodeproj
./Demos/DispatchQueueLogger/DispatchQueueLogger.xcodeproj
./Demos/LogFileCompressor/LogFileCompressor.xcodeproj
./Demos/CustomFormatters/CustomFormatters.xcodeproj
./Demos/GlobalLogLevel/GlobalLogLevel.xcodeproj
./Demos/CLI/CLI.xcodeproj
./Demos/PerUserLogLevels/PerUserLogLevels.xcodeproj
./Demos/WebServerIPhone/WebServerIPhone.xcodeproj
./Demos/UniversalApp/UniversalApp.xcodeproj
./Demos/OverflowTestMac/OverflowTestMac.xcodeproj
./Demos/CaptureASL/CaptureASL.xcodeproj
./Demos/CoreDataLogger/CoreDataLogger.xcodeproj
./Demos/TestXcodeColors/Desktop/TestXcodeColors.xcodeproj
./Demos/TestXcodeColors/Mobile/TextXcodeColors.xcodeproj
./Demos/RollingTestMac/RollingTestMac.xcodeproj
./Demos/NonArcTest/NonArcTest.xcodeproj
./Demos/RegisteredDynamicLogging/Desktop/RegisteredLoggingTest.xcodeproj
./Demos/RegisteredDynamicLogging/Mobile/RegisteredLoggingTest.xcodeproj
```
The cleanup will affect all of those projects.
Also, it will be super easy to change a setting in one place for all the repos. For example when upgrading the `SWIFT_VERSION`, one will just go to the `Module-Shared.xcconfig` and change it, all the projects picking that up.

**Comment by ffried** [comment]
I think we should also consider removing some outdated demos (like the NonArc one).

**Comment by bpoplauschi** [comment]
@ffried I really enjoy this talk :)

> But still the question is: Why don't we merge these into one project and use targets instead?
> Or at least merge all demos into one project. And tests and frameworks into one.

I think that is a good idea. Will do.

> There's only one risk, though. As soon as someone does not know about the xcconfig files and changes a setting directly on the target, the xcconfig file has lost and this setting is overwritten by the one in the project file. But I guess that's our responsibility to look out for this when reviewing changes.

I want to begin using [Danger](https://github.com/danger/danger) for our project and that will allow to auto-check certain things for each PR. For example, give a warning if xcodeproj settings were changed.

> I think we should also consider removing some outdated demos (like the NonArc one).

Agree about the NonArc.

**Comment by ffried** [comment]
@bpoplauschi 
> I think that is a good idea. Will do.

👍 

> I want to begin using Danger for our project and that will allow to auto-check certain things for each PR. For example, give a warning if xcodeproj settings were changed.

Heard a lot good about Danger, yea. Would be great if we can teach it to look out for build settings changes but accept changes like adding files or updating the project file version.
If you need any support setting it up, I'd be glad to help and learn about Danger, too. 🙂 

> Agree about the NonArc.

👍

**Comment by bpoplauschi** [comment]
@ffried created #956 - just a draft of ideas

**Comment by sushichop** [comment]
Thanks for your idea.

> I'm not why we would remove the other subspecs. But I agree CocoaPods subspecs are not 1:1 correspondents with Carthage or the manual installation.
> We can easily merge the Extensions into the Core subspec.
> Not sure about CLI since that only works on macos. But we already do it for Carthage so it should be fine.

I had thought I would make CocoaPods correspond to Carthage(1 to 1). 
But CLI subspec is actually used only by macOS, you say. So I'll left CLI subspec.
To summarize, subspecs will be Core(including Extension) aka Default, CLI, and Swift.

In addition, I'll fix CocoaLumberjack.modulemap file because it is different from the one created by CocoaPods. modulemap file must be considered when we support SwiftPM anyway.

I'll create a PR at the end of this week(Sadly, I don't have enough time today and tomorrow...😅).

> Also, as @ffried pointed, I did start to do do this merge - see the [xcode_10_merged_targets_use_xcconfig branch](https://github.com/CocoaLumberjack/CocoaLumberjack/tree/xcode_10_merged_targets_use_xcconfig)

Thanks. I confirmed.

P.S.
I have several other projects that use universal framework. 
And some project use xconfig and others don't use it.
(I feel it depends on the preference of the project members☺️)

**Comment by sushichop** [comment]
> And tests and frameworks into one.

I also agree.

**Comment by bpoplauschi** [comment]
@sushichop maybe you should wait until I am done with [xcode_10_merged_targets_use_xcconfig branch](https://github.com/CocoaLumberjack/CocoaLumberjack/tree/xcode_10_merged_targets_use_xcconfig) - I will try to work on it until the weekend ... keep in touch :)

**Comment by sushichop** [comment]
@bpoplauschi 
OK. I'll wait for it. Thanks.

**Comment by ffried** [comment]
> In addition, I'll fix CocoaLumberjack.modulemap file because it is different from the one created by CocoaPods. modulemap file must be considered when we support SwiftPM anyway.

Why do we have to manually create a module map? This is usually created automatically by the compiler / Xcode, right? Also, I don't think we need that for SwiftPM...

**Comment by sushichop** [comment]
@ffried 
Sorry, I may have misled you... We don't need to create a module map. It is automatically created by CocoaPods. It is just used by CocoaLumberjack target now.

About SwiftPM,  I think it will be needed if we support both Objective-C and Swift based target. It is not always need. See the good reference as following. I actually have the similar other project(private repo at the Github Enterprise).
https://github.com/google/promises

**Comment by ffried** [comment]
@sushichop 

> Sorry, I may have misled you... We don't need to create a module map. It is automatically created by CocoaPods. It is just used by CocoaLumberjack target now and Carthage integration.

No worries and thanks for the clarification!

> About SwiftPM, I think it will be needed if we support both Objective-C and Swift based target. It is not always needed.

Looking at what [Ankit wrote in his blog](http://ankit.im/swift/2016/05/21/creating-objc-cpp-packages-with-swift-package-manager/) about ObjC support in Swift PM, I don't think we need one. And IMHO it would be better if we don't needed one, since A) they're format is not final (yet) AFAIK, and B) it is again a easy target for problems if e.g. someone forgets to add a header file.
But we'll see once we're ready to add SwiftPM support I guess. 🙂

**Comment by bpoplauschi** [comment]
@sushichop you were correct the first time: we do have a manual modulemap: https://github.com/CocoaLumberjack/CocoaLumberjack/blob/master/Framework/Lumberjack/CocoaLumberjack.modulemap
CC @ffried

**Comment by sushichop** [comment]
@ffried 

> Looking at what [Ankit wrote in his blog](http://ankit.im/swift/2016/05/21/creating-objc-cpp-packages-with-swift-package-manager/) about ObjC support in Swift PM, I don't think we need one. And IMHO it would be better if we don't needed one

All right. I agree with you. Then let's support only Swift😁

@bpoplauschi 

> @sushichop you were correct the first time: we do have a manual modulemap: https://github.com/CocoaLumberjack/CocoaLumberjack/blob/master/Framework/Lumberjack/CocoaLumberjack.modulemap
> I don't know why we need it :)

This module map is basically used for Cocoalumberjack target when you integrate with Carthage.

For example, when if you use `DDDispatchQueueLogFormatter.h` with Cocoapods, it is OK by specifying `import CocoaLumberjack`. 
On the other hand, when if you use `DDDispatchQueueLogFormatter.h` with Carthage, you need to specify `import CocoaLumberjack.DDDispatchQueueLogFormatter`. 

I think this difference makes developers confused. Actually I have was confused.
In order to fix this(correspond to as possible),  I think it had better change module map file.

**Comment by sushichop** [comment]
I have rewrited above.
Sorry guys! I'll go to bed... It is midnight here(Japan)😪

**Comment by ffried** [comment]
@sushichop 
> I just have considered about the following.
See 0038-swiftpm-c-language-targets.md

You mean this part:
>The package manager will (a) construct a synthesized module map including all of the exported API headers, for use in Swift, and (b) will provide a header search path to this directory to any targets which transitively depend on this one.
>
>Module maps will only be synthesized for targets which either have a completely flat header layout (e.g., src/foo/include/*.h) or a single subdirectory (e.g., src/foo/include/foo/**.h). Any other structure requires the library author to provide explicit module maps. We may revisit this as we gain practical experience with it.

Our (current) structure would allow the module map file to be generated. And I think maintaining such a structure is probably easier than maintaining a module map ourselves. 🙂 

> Sorry guys! I'll go to bed... It is midnight here(Japan)😪

Have a good night!

**Comment by ksuther** [comment]
One suggestion about the minimum deployment version: I'd vote for watchOS 3.0. Targeting watchOS 3 or later is required to submit any watchOS app built for arm64_32, so target 2.0 will require that everyone either manually override the deployment version using CocoaPods or create a separate fork of lumberjack if not using CocoaPods (which is what we do).

**Comment by sushichop** [comment]
@ffried 

> Our (current) structure would allow the module map file to be generated. 

Thanks for sharing your idea. Until you heard your opinion, I convinced myself that the latest SwiftPM specification always needed a module map... 
But now, finally I understood that it is not always necessary.

> And I think maintaining such a structure is probably easier than maintaining a module map ourselves. 🙂

Yes. I also prefer this way now! I am really glad to discuss this😁
Thanks again!

**Comment by sushichop** [comment]
> One suggestion about the minimum deployment version: I'd vote for watchOS 3.0.

Thanks for your suggestion. I'd vote for watchOS 3.0 too.
Any idea? >all

**Comment by lolgear** [comment]
One more suggestion.

Lumberjack.xcodeproj has a lot of "deployment app targets" to test frameworks targets ( not test targets, hah ).

If they are necessary, it would be nice to put them in separate .xcodeproj as "DeploymentTestTargets.xcodeproj" and put them in "Integration" or "Tests" subdirectory.

And after that there is only two (three?) targets in main Lumberjack.xcodeproj:

1. CocoaLumberjack ( for iOS )
2. CocoaLumberjackSwift (for Swift)
3. (optional) CocoaLumberjackStatic ( static lib )

**Comment by bpoplauschi** [comment]
We have cleaned up the project significantly, but there are still some items remaining here.
From what I could capture from the thread above:
- [ ] `watchos` 3.0 deployment target
- Swift package manager - tracked with #869 
- [ ] module map - auto generate ?
- [ ] reduce the subspecs to 2 (core and swift) and match Carthage

Feel free to add if I missed anything

**Comment by sushichop** [comment]
@bpoplauschi 
Thanks for summary. I created the PR about subspecs.

**Comment by bpoplauschi** [comment]
@sushichop @ffried I think the warning started appearing after #959 

The cause is probably: `#if __has_include("CLIColor.h") && TARGET_OS_OSX`. 
We should just remove the `&& TARGET_OS_OSX`.
Basically always import the header and if the target is macos, then the file will actually have content.

For consistency, my opinion is to only have 2 subspecs and match the Carthage options:
- subspec `CocoaLumberjack/Core` matches exactly the Carthage `CocoaLumberjack.framework`
- subspec `CocoaLumberjack/Swift` matches exactly the Carthage `CocoaLumberjackSwift.framework`

Details: In #959 , I did put everything inside `CLIColor.{h,m}` between `#if TARGET_OS_OSX` and `#endif` because I did add it to the `CocoaLumberjack` unified target. Before we had a dedicated target for each platform, so the macOS target did contain it but the others didn't.

**Comment by sushichop** [comment]
@ffried @bpoplauschi 

> @sushichop @ffried I think the warning started appearing after #959
> 
> The cause is probably: `#if __has_include("CLIColor.h") && TARGET_OS_OSX`.
> We should just remove the `&& TARGET_OS_OSX`.
> Basically always import the header and if the target is macos, then the file will actually have content.

Yes, you are right. In more detailed, now we almost prevent importing `CLI.h` at `CocoaLumberjack.h` and `CLI.h`, but customized modulemap still exports `CLI.h` **implicitly**. That's why its warning is displayed.
https://github.com/CocoaLumberjack/CocoaLumberjack/blob/master/Framework/Lumberjack/CocoaLumberjack.modulemap

I don't know the historical reason why customized modulemap is adopted(because it had already been adopted when I joined CocoaLumberjack😝). But I assume developer adopted it in order to make it easy to prevent importing duplicated MACROS in those days. 
Then, now I think we don't need it. Because now CocoaLumberjack's headers prevent importing duplicated MACROS(by using #ifndef for example), though we need to confirm it. I haven't confirmed it  yet.

> For consistency, my opinion is to only have 2 subspecs and match the Carthage options:
> 
> * subspec `CocoaLumberjack/Core` matches exactly the Carthage `CocoaLumberjack.framework`
> * subspec `CocoaLumberjack/Swift` matches exactly the Carthage `CocoaLumberjackSwift.framework`

Yeah. I agree with you absolutely. I have been thinking this for a long time.Now it's time😃
Then I'll close #974 and create the other PR in order to reduce podspec to 2 subspecs.

**Comment by ffried** [comment]
Yea, like I already mentioned previously in another PR/issue, if anyhow possible I'd like to remove the custom module map and let Xcode do its job.

**Comment by sushichop** [comment]
@ffried @bpoplauschi 
I did create the PR in order to reduce podspec to 2subspecs and remove customized modulemap😃
See #976 .

**Comment by lolgear** [comment]
> @sushichop @ffried I think the warning started appearing after #959
> 
> The cause is probably: `#if __has_include("CLIColor.h") && TARGET_OS_OSX`.
> We should just remove the `&& TARGET_OS_OSX`.
> Basically always import the header and if the target is macos, then the file will actually have content.
> 
> For consistency, my opinion is to only have 2 subspecs and match the Carthage options:
> 
> * subspec `CocoaLumberjack/Core` matches exactly the Carthage `CocoaLumberjack.framework`
> * subspec `CocoaLumberjack/Swift` matches exactly the Carthage `CocoaLumberjackSwift.framework`
> 
> Details: In #959 , I did put everything inside `CLIColor.{h,m}` between `#if TARGET_OS_OSX` and `#endif` because I did add it to the `CocoaLumberjack` unified target. Before we had a dedicated target for each platform, so the macOS target did contain it but the others didn't.

This file `CLIColor.h` contains only a definition of structure Color.
Also, this file is imported only in one logger ( `DDTTYLogger` if I remember correctly ).

I suggest to put its definition and implementation into `DDTTYLogger`.

**Comment by bpoplauschi** [comment]
Since #981 was merged, I think we can close this issue. Right?

**Comment by lolgear** [comment]
@bpoplauschi 
I suggest to put `CLIColor` into `TTYLogger`.
Could it be considered and merged?

**Comment by bpoplauschi** [comment]
I don't know. I wouldn't change it without a good benefit and I don't see one. 
What do other people think? @CocoaLumberjack/collaborators

**Comment by ffried** [comment]
I'm with @bpoplauschi on this one. I don't see any benefit - especially since TTYLogger could probably be deprecated soon in favor of OSLogger.

**Comment by bpoplauschi** [comment]
@lolgear we can clean that up in our next major release.
